### ISSUE TITLE
hint should only regularize the coarse model

### ISSUE THREAD

**Issue by mikowals** [beopst/dcn.tf#1]
Thanks a lot for sharing this code.  I notice in the [objective-hint](https://github.com/beopst/dcn.tf/blob/master/models/dcn.py#L182) you are applying the loss to both coarse and fine features.  I think you want:

```
losses.l2_loss(flat_coarse - tf.stop_gradient(flat_fine), scope='objective_hint')
```

Otherwise you will train the fine model features to be more like the coarse features, which is not intended.

Do you mind sharing what error rate you have gotten down to on the validation set?  After 15 epochs I have a range of 16-17%.

**Comment by beopst** [comment]
@mikowals  Thank you for the comment! I just realized that I read the older version of the paper, and fixed the "objective_hint" loss as you suggested. 

After 15 epochs, I got the error rate on validation set around 9~11%.  I will try to reproduce the experimental results on the paper for cluttered MNIST dataset, and it will be updated soon.

**Comment by mikowals** [comment]
looks good.

**Comment by mikowals** [comment]
While I am thinking about it, the last line of section 6.1 in the paper (the appendix on mnist network for experiments) says that the coarse losses were used in convex combination.  Do you think that might mean there was a weight factor applied to the L2_loss before combining with the cross_entropy?  

Maybe something to consider if you have trouble getting down to the 1.3% error in the paper.

**Comment by beopst** [comment]
I think the meaning of "convex combination" is unclear. The authors may consider a trainable parameter \alpha to determine the relative importance between two terms. At this time, I am trying to consider a constant weight factor.

**Comment by mikowals** [comment]
I also just realized you probably don't want to update the trainable weights of the final layers based on the hint either.  To achieve that I think would require a separate optimizer that is fed with only the coarse layer parameters and hint loss.

**Comment by beopst** [comment]
I agree. I think it is better to use separate gradients without tf.stop_gradient ops. Thank you for your comment!
